vault/resource_token.go

```diff
@@ -0,0 +1,301 @@
+				ForceNew:    true,
+				Default:     "token",
+				Description: "The display name of the token.",
+			},
+			"num_uses": {
+				Type:        schema.TypeInt,
+				Required:    false,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The number of allowed uses of the token.",
+			},
+			"period": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "The period of the token.",
+			},
+			"lease_duration": {
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "The token lease duration.",
+			},
+			"lease_started": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The token lease started on.",
+			},
+			"accessor": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The client token accessor.",
+			},
+			"client_token": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "The client token.",
+				Sensitive:   true,
+			},
+		},
+	}
+}
+
+func tokenCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+
+	role := d.Get("role_name").(string)
+
+	iPolicies := d.Get("policies").(*schema.Set).List()
+	policies := make([]string, 0, len(iPolicies))
+	for _, iPolicy := range iPolicies {
+		policies = append(policies, iPolicy.(string))
+	}
+
+	var createRequest = &api.TokenCreateRequest{}
+
+	if len(policies) > 0 {
+		createRequest.Policies = policies
+	}
+
+	if v, ok := d.GetOk("ttl"); ok {
+		createRequest.TTL = v.(string)
+	}
+
+	if v, ok := d.GetOk("explicit_max_ttl"); ok {
+		createRequest.ExplicitMaxTTL = v.(string)
+	}
+
+	if v, ok := d.GetOk("period"); ok {
+		createRequest.Period = v.(string)
+	}
+
+	if v, ok := d.GetOk("no_parent"); ok {
+		createRequest.NoParent = v.(bool)
+	}
+
+	if v, ok := d.GetOk("no_default_policy"); ok {
+		createRequest.NoDefaultPolicy = v.(bool)
+	}
+
+	if v, ok := d.GetOk("display_name"); ok {
+		createRequest.DisplayName = v.(string)
+	}
+
+	if v, ok := d.GetOk("num_uses"); ok {
+		createRequest.NumUses = v.(int)
+	}
+
+	if v, ok := d.GetOk("renewable"); ok {
+		renewable := v.(bool)
+		createRequest.Renewable = &renewable
+	}
+
+	var resp *api.Secret
+	var err error
+
+	if role != "" {
+		log.Printf("[DEBUG] Creating token with role %q", role)
+		resp, err = client.Auth().Token().CreateWithRole(createRequest, role)
+		if err != nil {
+			return fmt.Errorf("error creating token with role %q: %s", role, err)
+		}
+
+		log.Printf("[DEBUG] Created token %q with role %q", resp.Auth.Accessor, role)
+	} else {
+		log.Printf("[DEBUG] Creating token")
+		resp, err = client.Auth().Token().Create(createRequest)
+		if err != nil {
+			return fmt.Errorf("error creating token: %s", err)
+		}
+
+		log.Printf("[DEBUG] Created token %q", resp.Auth.Accessor)
+	}
+
+	d.Set("lease_duration", resp.Auth.LeaseDuration)
+	d.Set("lease_started", time.Now().Format(time.RFC3339))
```

Comment:
It would be nice to get some test coverage on the `lease_duration` and `lease_started` fields because I've seen a couple of both time and duration-parsing bugs in the Terraform Vault Provider, and the test coverage would add a check that it really works, and an example of what kind of values to expect.

Reply:
Please check the two tests added lastly.

Reply:
This is amazing! @tyrannosaurus-becks could you please check? 🙏We would really appreciate this merged :)

Reply:
is this resolved?